### PR TITLE
Mappings: Remove ability to set path for _id and _routing on 2.0+ indexes

### LINKED ISSUE #6730: Deprecate extracting _routing and _id from document fields
Currently routing and ID values can be passed in the query string and url respectively, but they can also be extracted from fields within the document.

This has a significant performance impact because each doc needs to be parsed on the node which receives the index/get/update/delete request in order to know which node should handle the request.

On top of that, there are clashes between (eg) routing values set in fields and parent settings.

We should deprecate the functionality to extract these values from fields, and make it the responsibility of the client code instead.

It should still be possible to set `_routing` to required.  Perhaps we should set this automatically if the user ever passes in a routing or parent value at index time?

Relates to #8870

### PR DESCRIPTION
**Pull request by rjernst**
closes #6730

Note: I would like to keep the BulkTests checking pre-parsing when path is set works, but I don't see a way without moving that to the static backcompat tests.  It would be nice if we could fake out the index settings (ie set the index_created_version) but I don't think that is possible...

### CODE REVIEW

**Comment by martijnvg** [comment]
@rjernst I like the cleanup! If I recall correctly, setting the version is a way of the fake the index version. So maybe try this out?

**Comment by rjernst** [comment]
@martijnvg That worked! I've added back a couple of the tests so there is some direct testing of backcompat here.

File: src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java [hunk 1]
```diff
@@ -144,6 +146,8 @@ protected IdFieldMapper(String name, String indexName, float boost, FieldType fi
         super(new Names(name, indexName, indexName, name), boost, fieldType, docValues, Lucene.KEYWORD_ANALYZER,
                 Lucene.KEYWORD_ANALYZER, postingsProvider, docValuesProvider, null, null, fieldDataSettings, indexSettings);
         this.path = path;
+        this.writePre20Settings = indexSettings.equals(ImmutableSettings.EMPTY) == false &&
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java]: Just wondering, but I don't think that the `indexSettings.equals(ImmutableSettings.EMPTY) == false` part of the expression is needed?

Comment (rjernst) [inline src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java]: It is, because `Version.indexCreated` throws an exception if it cannot find the version setting (and the default constructor above uses empty settings).

Comment (martijnvg) [inline src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java]: Right I see... I think the `Version.indexCreated` should just return null in that case, but that is outside of the scope of this change.

**Comment by martijnvg** [comment]
LGTM!

File: src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java (hunk 1 above)
Comment (jpountz) [inline src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java]: @rjernst It looks like we can easily make this constructor take some index settings. It is only called from DocumentMapper.Builder which has index settings available (btw some other mappers already take settings).

File: src/test/java/org/elasticsearch/routing/SimpleRoutingTests.java [hunk 2]
```diff
@@ -331,8 +359,7 @@ public void testRequiredRoutingWithPathNumericType() throws Exception {
             assertThat(client().prepareGet(indexOrAlias(), "type1", "1").setRouting("0").execute().actionGet().isExists(), equalTo(true));
         }
     }
-
-    @Test
```
Comment (jpountz) [inline src/test/java/org/elasticsearch/routing/SimpleRoutingTests.java]: :)

**Comment by jpountz** [comment]
LGTM too but can you please fix the `indexSettings.equals(ImmutableSettings.EMPTY)` hack before pushing?

File: src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java (hunk 1 above)
Comment (martijnvg) [inline src/main/java/org/elasticsearch/index/mapper/internal/IdFieldMapper.java]: Since all 1.x versions will have the created version in the index settings, the empty settings check can be removed.
